### DISCUSSION TITLE
Why run external_converters in vm?

### DISCUSSION THREAD

**Discussion by lovelylain** [Koenkk/zigbee2mqtt#5307] [General]
At 153b82ca7735e3c9c8f3226c561e04ada71149f9 external_converters.js was changed to run in VM, which was not easy for me.
Before that, I can solve the `Cannot find module` problem in this way:
```
const path = require.main.path
module.paths.push(path + '/node_modules')
```
And I can use console.log to debug, now console.log not work.
Also I can improve the function by patching the existing devices, there is a problem after introducing VM.
```
fz_apply('QBKG12LM', ['QBKG24LM'],
  'xiaomi_on_off_action',
  'xiaomi_multistate_action',
  'xiaomi_operation_mode_basic',
)
tz_apply('QBKG12LM', ['QBKG24LM'],
  'xiaomi_switch_operation_mode',
)
```
Can you change it back?

**Comment by Koenkk** [comment]
@nurikk If I understand correctly we cannot change this back because it will break the HA setup, right?

**Reply by nurikk** [reply]
Yes, this will break HA setup. If you need console, then you have to add this global available here (currently only require function available globally) :  https://github.com/Koenkk/zigbee2mqtt/blob/05142044cbe3e6f9379914bf5f19d2a3d12622d8/lib/util/utils.js#L154

adding console to external converters will look something like this:

```javascript
            const sandbox = {
                require,
                console,
                module: {},
            };
```

**Reply by lovelylain** [reply]
I notice the `else` condition:
```js
        if (moduleName.endsWith('.js')) {
            // ...
            vm.runInNewContext(converterCode, sandbox);
            converter = sandbox.module.exports;
        } else {
            converter = require(moduleName);
        }
```
what will happen if I strip `.js` suffix:
```yaml
external_converters:
  #- custom_devices.js
  - custom_devices
```

**Reply by lovelylain** [reply]
It will result in `Cannot find module`, can you add an option to not run in vm?

**Reply by nurikk** [reply]
yes, you can register your code as npm package and load in this way more about npm packages [here](https://docs.npmjs.com/creating-node-js-modules)

**Reply by lovelylain** [reply]
```yaml
data_path: /share/zigbee2mqtt
external_converters:
  - /share/zigbee2mqtt/custom_devices
```
It works if I write absolute path, no need to create package.
Problem solved, thanks~

**Reply by nurikk** [reply]
Definitely not intended behavior, hacky solution
